### PR TITLE
Make GrMhd system compatible with 3D EoS

### PR DESCRIPTION
**Pull request by isaaclegred**
## Proposed changes

This PR adds support for 3D thermodynamic EoSs in the evolution system.  Very little is changed (another instantiation is added) except in primitive recovery, where some logic needed to be added.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by isaaclegred** [comment]
@nilsvu, @nilsdeppe, (perhaps @ermost should take a look at the modifications to primitive recovery too)

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp
Comment (nilsvu) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp:294]: Evaluate only the EOS bounds in the `if` and then compute epsilon outside the `if` to make clear that the expression for epsilon is independent of the thermodynamic dim.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveFromConservative.cpp
Comment (nilsvu) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveFromConservative.cpp]: `else`

Also this file needs to be run through clang-format. If you do, please commit separately.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp
Comment (nilsdeppe) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp:294]: No, that spills scope and adds extra complexity.

I would do:
```cpp
double epsilon_hat = w_hat * (q_bar - mu * r_bar_squared) +
                       v_hat_squared * square(w_hat) / (1.0 + w_hat);
if constexpr ((ThermodynamicDim == 3) {
    epsilon_hat =
        std::clamp(epsilon_hat,
                   equation_of_state_.specific_internal_energy_lower_bound(
                       rho_hat, electron_fraction_),
                   equation_of_state_.specific_internal_energy_upper_bound(
                       rho_hat, electron_fraction_));
} else {
    epsilon_hat = std::clamp(
       epsilon_hat,
        equation_of_state_.specific_internal_energy_lower_bound(rho_hat),
        equation_of_state_.specific_internal_energy_upper_bound(rho_hat));
  }

```

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: `const Scalar<double>`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.cpp]: Oh, CI is complaining about a missing semicolon on this line :)

**Comment by ermost** [comment]
Hi @isaaclegred,

this looks great! The Con2Prim changes in Newman and KastaunMHD look good. Do you also want to modify KastaunHydro?
(I think you could just copy the code you changed in the MHD routine.)

**Comment by isaaclegred** [comment]
> this looks great! The Con2Prim changes in Newman and KastaunMHD look good. Do you also want to modify KastaunHydro? (I think you could just copy the code you changed in the MHD routine.)

Yep I can definitely do this, @ermost!

**Comment by isaaclegred** [comment]
@nilsdeppe should I rebase on develop and fix the conflict?

**Comment by nilsdeppe** [comment]
Yes, though maybe wait a bit because #5494 will probably also cause a conflict?

Edit: I'm currently rebasing that one

**Comment by isaaclegred** [comment]
@nilsdeppe and @nilsvu any other comments or should I squash?

**Review by nilsdeppe — COMMENTED** [review]
Looks good, please squash!

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp
Comment (nilsvu) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp]: Please fix the formatting here when squashing. Just squash, then run `git-clang-format HEAD~` and squash in the changes from that too.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAlHydro.cpp
Comment (nilsvu) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAlHydro.cpp:235]: This shouldn't compile for dim 3 because `specific_internal_energy_lower_bound` takes 2 arguments, right?

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/PalenzuelaEtAl.cpp
Comment (nilsvu) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/PalenzuelaEtAl.cpp]: clang-tidy rightfully complains here that `specific_internal_energy` already exists above.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveFromConservative.cpp
Comment (nilsvu) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveFromConservative.cpp]: You're missing `specific_internal_energy` (4th argument)

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp
Comment (nilsvu) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp:317]: You can ignore the "c-style casts are discouraged" warnings from clang-tidy (we'll have to fix those for tensors like this).

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Comment by nilsvu** [comment]
And yes please squash

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp
Comment (isaaclegred) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp:317]: Do you mean literally ignore them or NOLINT them?  :)

Comment (nilsvu) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp:317]: Ignore, no need to nolint because we have to fix those more permanently (this pattern occurs all over the code, can't nolint it everywhere)

**Review by nilsvu — CHANGES_REQUESTED** [review]
Changes look good, but it looks like none of the ThermodynamicDim=3 code is tested. That seems important since we'll mostly run with 3D EOSs once the other changes to the executables go through. Could you please add tests for the code added here?

**Comment by ermost** [comment]
For the 3D Primitive recovery test, would it be possible to check what happens when the evolved electron fraction goes out of bounds?

**Comment by isaaclegred** [comment]
> For the 3D Primitive recovery test, would it be possible to check what happens when the evolved electron fraction goes out of bounds?

@ermost I can certainly add a test,  what is the expected behavior in that case?

**Comment by ermost** [comment]
That would be great, thank you @isaaclegred!

It should be pushed back to the bounds of the EOS. so if you input ye_star < 0, you should get ye_min, and if you have ye_star > rho*ye_max, you should get ye_max.

The undertermined part is what happens if D <0, but that should be caught in the fix atmosphere, or fix conservative part. i.e., whether you keep ye, or just replace it by the atmospheric value.

**Comment by isaaclegred** [comment]
@ermost, @nilsdeppe, @nilsvu  whenever you get a chance this is ready for review! I've also added some potentially unnecessary tests to GhMhd subcell related things with 3d EoSs.  Let me know if there's anything else you'd like to see.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp:133]: LGTM, this has a catastrophic cancellation, but I don't know what to do about that other than to rework the entire recovery scheme (not worth it!)

**Review by nilsdeppe — COMMENTED** [review]
I'm happy for you to squash once @ermost has taken a look :)

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (isaaclegred) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp:133]: Yeah, I would guess we would be using Kastaun most of the time anyway, if this becomes a serious limitation we  can take a look at trying to patch it up.

Comment (ermost) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp:133]: I think it's an intrinsic limitation of the Newman Hamlin scheme in terms of iterating over pressure. For using tabulated EOS the above has always been fine for me, when I was using Newman-Hamlin.

Comment (nilsdeppe) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp:133]: Yea, I think the "solution" is "use a better recovery scheme" 😅 Kastaun isn't perfect in this regard, but it isn't obvious to me that a "perfect" scheme is even possible

Comment (ermost) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp:133]: I think something better would probably go towards a physicality preserving evolution scheme, but that is even harder :)

**Comment by ermost** [comment]
Thanks a lot for additional test @isaaclegred . This looks good to me.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp
Comment (nilsdeppe) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp:133]: Yes, that would be the dream... But agreed, very, very difficult :(

**Comment by isaaclegred** [comment]
Just rebased on develop to fix conflicts

**Comment by nilsdeppe** [comment]
Uh oh, something went wrong during the rebase :(

**Comment by isaaclegred** [comment]
Yeah Con2Prim is broken now (like after fixing things to make it compile), I'm investigating

**Comment by isaaclegred** [comment]
@nilsdeppe @ermost  I think this is ready to be looked at again.

**Comment by ermost** [comment]
Looks good! Thanks for fixing this!

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
LGTM, please squash :D

**Comment by isaaclegred** [comment]
Changed some things to get rid of clang-tidy complaints, should be good to go when CI completes @nilsdeppe and @ermost!

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp:316]: If you wanted to fix these, I think you need to swap the parens for braces: `Scalar<double>{...}`

**Review by nilsdeppe — DISMISSED** [review]
@nilsvu please take a look again...

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp
Comment (isaaclegred) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp:316]: @nilsvu said to ignore these earlier I think, which is why I didn't change them, but I can change them if that's the way things are going.

Comment (nilsdeppe) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp:316]: Sounds good!

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp
Comment (nilsvu) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp]: (optional) move this code outside the `if` because it's repeated in both branches?

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp
Comment (nilsvu) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp]: (optional) Move this line outside the `if` as well, and only do `q_ = std::max(q_, eps_min)` in the `EnforcePhysicality` branch? Seems easier to understand that way.

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_PrimsAfterRollback.cpp
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_PrimsAfterRollback.cpp]: I recommend you build clang-format into your workflow, it makes editing code a lot easier. Try running `git-clang-format` and/or use your editor's formatting extensions.

**Review by nilsvu — DISMISSED** [review]
Merging as-is is fine, or squash in these things if you want to do them and merge then.

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_PrimsAfterRollback.cpp
Comment (isaaclegred) [reply tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_PrimsAfterRollback.cpp]: clang-format for some reason doesn't remove this whitespace  :(

**Comment by isaaclegred** [comment]
@nilsvu, @nilsdeppe changes fixed and squashed

**Review by nilsdeppe — APPROVED** [review]

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_PrimsAfterRollback.cpp
Comment (nilsvu) [reply tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_PrimsAfterRollback.cpp]: Maybe the extra empty line tells it not to reformat

**Review by nilsvu — APPROVED** [review]

**Pull request merged**
